File: shoring/desk/src/main/java/com/panopset/marin/oldswpkg/games/blackjack/FxChipPainter.java
package com.panopset.marin.oldswpkg.games.blackjack;

import java.util.ArrayList;
import java.util.List;

import com.panopset.fxapp.FontManagerFX;
import com.panopset.marin.game.card.images.BlackjackImages;
import com.panopset.marin.oldswpkg.cards.pngmap.Chip;
import com.panopset.marin.oldswpkg.cards.pngmap.ChipPainter;

import javafx.scene.canvas.GraphicsContext;
import javafx.scene.image.Image;

public class FxChipPainter {

  public FxChipPainter() {
    Double cw = getChipImg().getWidth();
    Double ch = getChipImg().getHeight();
    cp = new ChipPainter(cw.intValue(), ch.intValue());
  }

  public int getChipHeight() {
    return cp.getChipHeight();
  }

  public int getChipWidth() {
    return cp.getChipWidth();
  }

  private Image chipImg;

  private Image getChipImg() {
    if (chipImg == null)  {
        chipImg = BlackjackImages.getChipMap();
    }
    return chipImg;
  }


  public int paintChips(GraphicsContext g, int x, int y, long amt) {
    List<Chip> stack = new ArrayList<Chip>();
    long newAmt = amt;
    WHILE: while (newAmt > 99) {
      FOR: for (Chip chip : Chip.values()) {
        if (chip.getValue() == 250 && newAmt > 250) {
          continue FOR;
        }
        if (newAmt >= chip.getValue()) {
          stack.add(chip);
          newAmt = newAmt - chip.getValue();
          continue WHILE;
        }
      }
    }
    int xloc = x + (2 * stack.size());
    int yloc = y;
    for (Chip chip : stack) {
      paintChip(chip, g, xloc, yloc);
      yloc = (int) (yloc - (FontManagerFX.getImgRatio() * 2));
      xloc = (int) (xloc - (FontManagerFX.getImgRatio() * 1));
    }
    return yloc;
  }

  private void paintChip(final Chip chip, final GraphicsContext g, final int x,
    final int y) {
    int[] b = cp.getPaintDimensionsFX(chip, x, y);
    g.drawImage(chipImg, b[0], b[1], b[2], b[3], b[4], b[5], b[6], b[7]);
  }

  private final ChipPainter cp;

}

File: code/shoring/desk/src/main/kotlin/com/panopset/marin/oldswpkg/games/blackjack/FxChipPainter.kt
package com.panopset.marin.oldswpkg.games.blackjack

import com.panopset.fxapp.FontManagerFX
import com.panopset.marin.game.card.images.BlackjackImages
import com.panopset.marin.oldswpkg.cards.pngmap.Chip
import com.panopset.marin.oldswpkg.cards.pngmap.ChipPainter
import javafx.scene.canvas.GraphicsContext
import javafx.scene.image.Image

class FxChipPainter {
    val chipHeight: Int
        get() = cp.chipHeight
    val chipWidth: Int
        get() = cp.chipWidth
    private var chipImg: Image? = null
    private fun getChipImg(): Image? {
        if (chipImg == null) {
            chipImg = BlackjackImages.chipMap
        }
        return chipImg
    }

    fun paintChips(g: GraphicsContext, x: Int, y: Int, amt: Long): Int {
        val stack: MutableList<Chip> = ArrayList()
        var newAmt = amt
        WHILE@ while (newAmt > 99) {
            FOR@ for (chip in Chip.entries) {
                if (chip.value == 250 && newAmt > 250) {
                    continue@FOR
                }
                if (newAmt >= chip.value) {
                    stack.add(chip)
                    newAmt -= chip.value
                    continue@WHILE
                }
            }
        }
        var xloc = x + 2 * stack.size
        var yloc = y
        for (chip in stack) {
            paintChip(chip, g, xloc, yloc)
            yloc = (yloc - FontManagerFX.imgRatio * 2).toInt()
            xloc = (xloc - FontManagerFX.imgRatio * 1).toInt()
        }
        return yloc
    }

    private fun paintChip(
        chip: Chip, g: GraphicsContext, x: Int,
        y: Int
    ) {
        val b = cp.getPaintDimensionsFX(chip, x, y)
        g.drawImage(
            chipImg,
            b[0].toDouble(),
            b[1].toDouble(),
            b[2].toDouble(),
            b[3].toDouble(),
            b[4].toDouble(),
            b[5].toDouble(),
            b[6].toDouble(),
            b[7].toDouble()
        )
    }

    private val cp: ChipPainter

    init {
        val cw = getChipImg()!!.width
        val ch = getChipImg()!!.height
        cp = ChipPainter(cw.toInt(), ch.toInt())
    }
}
